From 8a3185792a0d61c15091db66d970ea8dce591871 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Thu, 26 Jan 2017 13:56:18 +0200
Subject: [PATCH 1313/2241] fix: net: mvpp2x: fix traffic stuck due wrong
 stopped netdev_queue

Issue:
- In some traffic scenarios skb->sender_cpu different from netdev_queue
  higher bits -> wrong netdev_queue stopped and traffic stuck.

Fix:
- Stop netdev_queue only if skb->queue_mapping equal to txq->log_id
  on current CPU. netdev_queue would be started in TX done routine,
  netdev_queue ID based on txq->log_id and CPU ID.
  In xmit procedure: CPU ID = current CPU
  In Tx Done procedure: CPU ID = interrupt cause CPU ID

Change-Id: I7fb14810c7ac66bf5e5b742c4d9579f146bb95fd
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/36072
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index cba518c..719ace2 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -2881,7 +2881,7 @@ static int mv_pp2x_tx(struct sk_buff *skb, struct net_device *dev)
 
 	/* Prevent shadow_q override, stop tx_queue until tx_done is called*/
 	if (unlikely(mv_pp2x_txq_free_count(txq_pcpu) < port->txq_stop_limit)) {
-		if (cpu == skb->sender_cpu) {
+		if ((txq->log_id + (cpu * mv_pp2x_txq_number)) == skb_get_queue_mapping(skb)) {
 			nq = netdev_get_tx_queue(dev, skb_get_queue_mapping(skb));
 			netif_tx_stop_queue(nq);
 		}
-- 
2.7.4

